### ISSUE TITLE
Encountered an issue when trying to optimize Donut model.

### ISSUE THREAD

**Issue by dneemuth** [neuralmagic/deepsparse#1146]
Hi ,
 I've been experimenting with your excellent tool (awesome work) to try to optimize my Donut model, but currently struggling with a weird issue. I have posted the google collab link below, any help/assistance would be much appreciated. thanks.

https://colab.research.google.com/drive/1iBtD8Ceas2AWQLw8UcCsuQvkdL3JD4QV?usp=sharing

**Comment by anmarques** [comment]
Hi @dneemuth. Thanks for you interest in using DeepSparse to accelerate your inference.

I inspected your colab and it seems that you're trying to pass the image pixel values as input to "decoder_model.onnx". That's not the intended use of this architecture.

Donut is an encoder-decoder architecture, in which the image input is processed by the encoder. The output of the encoder (encoder_hidden_states) is then passed to the decoder. The decoder generates one token at time, until the maximum number of tokens is generated or an end-of-sentence token is generated. The decoder takes as input the encoder_hiden_states, input_ids (existing tokens up to that decoding step), and attention_mask.

To enable that with the DeepSparse engine you'll need to create two DeepSparse pipelines: one for the encoder and one for the decoder.

I hope this helps!

**Comment by dneemuth** [comment]
@anmarques thanks very much for your assistance and elaborate explanation. greatly appreciate. I've updated my collab notebook and have been able to make some decent progress by using your proposed solution e.g. by using 2 DeepSparse pipelines.
1. created a custom DeepSparse pipeline for Encoder to accept image input.
2. created a custom DeepSparse pipepline that takes as input  encoder_hidden_states, input_ids (existing tokens up to that decoding step), and attention_mask.

Only issue I'm getting now is trying to figure out how to process the output from the Custom Decoder Pipeline. if you have some time , can you please have a look and point me in the right direction. again thanks again for your invaluable support.

https://colab.research.google.com/drive/1iBtD8Ceas2AWQLw8UcCsuQvkdL3JD4QV?usp=sharing

**Comment by anmarques** [comment]
Hi @dneemuth. Good to see you're making progress.

First I must mention that what you did wrapping ONNXRuntime sessions inside the regular VisionEncoderDecoderModel class should also work for DeepSparse. If you create your own full pipeline you can do a cleaner job, but this might be a good starting point since it seems you made it work for ORT.

Going into more details, the decoder will output are the logits. If you use the VisionEncoderDecoderModel route above, the parent class will handle the output from that point onwards. I don't think you have to do any post-processing. On the other hand, if you try to do the post-processing yourself, you first need to transform the logits into a probability distribution using a softmax function. Then you sample from that distribution to generate tokens. Here's some more explanation on this process: [https://huggingface.co/blog/how-to-generate](https://huggingface.co/blog/how-to-generate).

As a final note, I noticed you're using read_and_crop_original in the pre-processing for the decoder. I believe that this function does the preprocessing on the image. This is what you need for the encoder input.

**Comment by NMfitz** [comment]
Hi @dneemuth apologies for piggy-backing on this ticket - I am the UX lead at NM and saw that you had filled out the model request form on SparseZoo. Wanted to reach out and ask if you had any observations (likes and dislikes) about the Zoo that you would be interested in sharing. Please let me know - thank you!

**Comment by NMfitz** [comment]
@dneemuth thank you so much for the great feedback! I've shared it with the SparseZoo team. The deply/contribute models to the Zoo is duly noted  (+1!)  Apologies for the lack of pricing info; I know we do have a [90-day free trial](https://neuralmagic.com/deepsparse-free-trial/) of our DeepSparse engine but LMK if I can help get you more info - thanks again - "Fitz"

**Comment by jeanniefinks** [comment]
Hello @dneemuth 
As there has been no further comments to this thread, I will go ahead and close it. Please re-open if you have additional questions, thanks!
Jeannie // Neural Magic
